util/reggen/gen_rust.py

```diff
@@ -0,0 +1,388 @@
+        sys.exit(1)
+
+    define_declare = 'pub const ' + name + ': ' + data_type(name, val)
+
+    oneline_define = define_declare + ' = ' + val + ';'
+
+    existing_defines.add(name)
+
+    return oneline_define + '\n'
+
+
+def gen_const_register(outstr: TextIO,
+                       reg: Register,
+                       comp: str,
+                       width: int,
+                       rnames: Set[str],
+                       existing_defines: Set[str]) -> None:
+    rname = reg.name
+    offset = reg.offset
+
+    genout(outstr, format_comment(first_line(reg.desc)))
+    defname = as_define(comp + '_' + rname)
+    statement = gen_const(defname, '_REG_OFFSET', hex(offset), existing_defines)
+    genout(outstr, statement)
+
+    for field in reg.fields:
+        dname = defname + '_' + as_define(field.name)
+        field_width = field.bits.width()
+
+        if field_width == 1:
+            # single bit
+            statement = gen_const(dname, '_BIT', str(field.bits.lsb), existing_defines)
+            genout(outstr, statement)
+        else:
+            # multiple bits (unless it is the whole register)
+            if field_width != width:
+                mask = field.bits.bitmask() >> field.bits.lsb
+                statement = gen_const(dname, '_MASK', hex(mask), existing_defines)
+                genout(outstr, statement)
+
+                statement = gen_const(dname, '_OFFSET', str(field.bits.lsb), existing_defines)
+                genout(outstr, statement)
+            if field.enum is not None:
+                for enum in field.enum:
+                    ename = as_define(enum.name)
+                    value = hex(enum.value)
+                    statement = gen_const(
+                        defname + '_' + as_define(field.name),
+                        '_VALUE_' + ename,
+                        value,
+                        existing_defines)
+                    genout(outstr, statement)
+    genout(outstr, '\n')
+    return
+
+
+def gen_const_window(outstr: TextIO,
+                     win: Window,
+                     comp: str,
+                     regwidth: int,
+                     rnames: Set[str],
+                     existing_defines: Set[str]) -> None:
+    offset = win.offset
+
+    genout(outstr, format_comment('Memory area: ' + first_line(win.desc)))
+    defname = as_define(comp + '_' + win.name)
+    genout(
+        outstr,
+        gen_const(defname, '_REG_OFFSET', hex(offset), existing_defines))
+    items = win.items
+    genout(
+        outstr,
+        gen_const(defname, '_SIZE_WORDS', str(items), existing_defines))
+    items = items * (regwidth // 8)
+    genout(
+        outstr,
+        gen_const(defname, '_SIZE_BYTES', str(items), existing_defines))
+
+    wid = win.validbits
+    if (wid != regwidth):
+        mask = (1 << wid) - 1
+        genout(outstr,
+               gen_const(defname, '_MASK ', hex(mask), existing_defines))
```

Comment:
Trailing space after MASK here (presumably doesn't matter because you just get an extra space in the output, but I think we should tidy it up).

Reply:
Done.